### PR TITLE
Source extensions

### LINKED ISSUE #1: Multiple source file extensions
We should support using multiple source file extensions at once, such as `.cpp` and `.cc`.
The user should be able to specify multiple extensions explicitly, or to specify a pattern for source file extensions or filenames.

For example:
`SRCEXT = .cpp .cc`
or:
`SRCPAT = "*.cpp|*.cc"`

### PR DESCRIPTION
**Pull request by Thyrum**
This pull would add support for multiple file extensions. Multiple exenstions can be provided using `SRCEXT ::= .cpp .cc`, that is, a space seperated list.

Closes #1

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -70,17 +70,27 @@ MKBIN_D = $(CXX) $(CXXFLAGS) $(FLAGS_D) $^ $(LDFLAGS) -o $@
...
-OBJS_R  ::= $(patsubst $(SRCDIR)/%$(SRCEXT),$(OBJDIR_R)/%.o,$(SOURCES))
-OBJS_D  ::= $(patsubst $(SRCDIR)/%$(SRCEXT),$(OBJDIR_D)/%.o,$(SOURCES))
+SOURCES ::= $(foreach ext, $(SRCEXT), $(wildcard $(SRCDIR)/*$(ext)))
+OBJS_R  ::= $(foreach ext, $(SRCEXT), $(patsubst $(SRCDIR)/%$(ext),$(OBJDIR_R)/%.o,$(filter %$(ext), $(SOURCES))))
+OBJS_D  ::= $(foreach ext, $(SRCEXT), $(patsubst $(SRCDIR)/%$(ext),$(OBJDIR_D)/%.o,$(filter %$(ext), $(SOURCES))))
 
-DEPS ::= $(patsubst $(SRCDIR)/%$(SRCEXT),$(DEPDIR_R)/%.d,$(SOURCES)) \
-         $(patsubst $(SRCDIR)/%$(SRCEXT),$(DEPDIR_D)/%.d,$(SOURCES))
+DEPS ::= $(foreach ext, $(SRCEXT), $(patsubst $(SRCDIR)/%$(ext),$(DEPDIR_R)/%.d,$(filter %$(ext), $(SOURCES)))) \
+         $(foreach ext, $(SRCEXT), $(patsubst $(SRCDIR)/%$(ext),$(DEPDIR_D)/%.d,$(filter %$(ext), $(SOURCES))))
```
Comment (avdstaaij) [inline Makefile:93]: These lines are now over 80 characters long, which is not preferred.
However, I do not see an easy way to shorten them, so it is acceptable.
If you know a way to shorten them that does not hinder readability, you are of course encouraged to do so.

**Review by avdstaaij — CHANGES_REQUESTED** [review]
I have tested the new Makefile and it works perfectly!
I have not tested the version check, but the implementation looks very clean and the source seems trustable.

In the additions, you have written all new variables in lowercase. All existing variables are in uppercase, so this is a bit of clash. I recommend making the new variables uppercase as well. The `compile_rule` is a very different type of object, so I think keeping that lowercase is fine. Since it encompasses two rules, it may be more accurate to call it `compile_rules`, but the singular form is fine as well.

If you are going to apply my proposed changes, I suggest you amend the first commit via interactive rebase to keep the history clean.

The readme still needs to be amended before merging. (FIXED)

The second commit title should read `>=` instead of `>`. (FIXED)

**Comment by Thyrum** [comment]
> In the additions, you have written all new variables in lowercase. All existing variables are in uppercase, so this is a bit of clash.

My reasoning for using lower case variable names was that these are "local"/"scoped" variables. I find it helpful to distinguish between the scope of variables in such a manner (I have also done this on other pull requests if I recall correctly).
The variable `has_features` however should definitely be upper case.

**Comment by avdstaaij** [comment]
> My reasoning for using lower case variable names was that these are "local"/"scoped" variables. I find it helpful to distinguish between the scope of variables in such a manner (I have also done this on other pull requests if I recall correctly).
> The variable `has_features` however should definitely be upper case.

That makes sense.

**Comment by avdstaaij** [comment]
I have added this PR's feature to the readme and fixed the `>`/`>=` mixup.

**Comment by avdstaaij** [comment]
This branch now needs to be rebased on `master`. @Thyrum, you would probably do this best.

**Comment by Thyrum** [comment]
As discussed in #1, the code complication is not worth the added functionality.

**Pull request closed without merging**